core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4061,11 +4074,40 @@ public int toDate(String dateString, String fmtString) {
           new java.sql.Date(internalToDateTime(dateString, fmtString)));
     }
 
+    public int toDatePg(String dateString, String fmtString) {
+      try {
+        return (int) PostgresqlDateTimeFormatter.toTimestamp(dateString, fmtString,
+                LOCAL_ZONE)
+            .getLong(ChronoField.EPOCH_DAY);
+      } catch (Exception e) {
+        SQLException sqlEx =
+            new SQLException(
+                String.format(Locale.ROOT,
+                    "Invalid format: '%s' for datetime string: '%s'.", fmtString,
+                    dateString));
+        throw Util.toUnchecked(sqlEx);
+      }
+    }
+
     public long toTimestamp(String timestampString, String fmtString) {
       return toLong(
           new java.sql.Timestamp(internalToDateTime(timestampString, fmtString)));
     }
 
+    public long toTimestampPg(String timestampString, String fmtString) {
+      try {
+        return PostgresqlDateTimeFormatter.toTimestamp(timestampString, fmtString, LOCAL_ZONE)
```

Comment:
Matching the behaviour of the existing `to_timestamp` method. It will call `toLong()`
https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java#L3743

which will always use the TimeZone `LOCAL_TZ`.

Reply:
Is this what Postgres expects?

Reply:
PostgreSQL will parse the timestamp in the default timezone unless the format and input have timezone information.

From my understanding, I need to eventually return a long value (millis since epoch in UTC) for a timestamp value. Is it possible to return a timestamp with timezone information?
https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java#L4064

Since I am returning a long value (millis since epoch in UTC), I am using UTC as the default timezone.

Reply:
I have added recently support for timestamps with time zone

Reply:
I don't see how to return a timestamp with a timezone. This may not be an issue though. If the timestamp format does not have timezone patterns, then it will be assumed to be `LOCAL_ZONE` (UTC). If the pattern has timezone patterns, then it is parsed using the timezone from the input string. After parsing, it is converted to `LOCAL_ZONE` (UTC).

As an example, if you have the input `2024-04-17 00:00:00-07:00` and pattern `YYYY-MM-DD HH24:MI:SSTZH:TZM`, you would end up with `2024-04-17 07:00:00`.